core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
+                !a <- decode
+                !b <- decode
+                !c <- decode
+                !d <- decode
+                return (a, b, c, d)
+
+instance Bi BS.ByteString where
+    encode = E.encodeBytes
+    decode = D.decodeBytes
+
+instance Bi Text.Text where
+    encode = E.encodeString
+    decode = D.decodeString
+
+instance Bi BS.Lazy.ByteString where
+    encode = encode . BS.Lazy.toStrict
+    decode = BS.Lazy.fromStrict <$> decode
+
+instance Bi a => Bi [a] where
+    encode = encodeList
+    decode = decodeList
+
+instance (Bi a, Bi b) => Bi (Either a b) where
+    encode (Left  x) = E.encodeListLen 2 <> E.encodeWord 0 <> encode x
+    encode (Right x) = E.encodeListLen 2 <> E.encodeWord 1 <> encode x
+
+    decode = do D.decodeListLenOf 2
+                t <- D.decodeWord
+                case t of
+                  0 -> do !x <- decode
+                          return (Left x)
+                  1 -> do !x <- decode
+                          return (Right x)
+                  _ -> fail $ "decode@Either: unknown tag " <> show t
+
+instance Bi a => Bi (NonEmpty a) where
+  encode = defaultEncodeList . toList
+  decode = do
+    l <- defaultDecodeList
+    case nonEmpty l of
+      Nothing -> fail "Expected a NonEmpty list, but an empty list was found!"
+      Just xs -> return xs
+
+instance Bi a => Bi (Maybe a) where
+    encode Nothing  = E.encodeListLen 0
+    encode (Just x) = E.encodeListLen 1 <> encode x
+
+    decode = do n <- D.decodeListLen
+                case n of
+                  0 -> return Nothing
+                  1 -> do !x <- decode
+                          return (Just x)
+                  _ -> fail $ "decode@Maybe: unknown tag " <> show n
+
+encodeContainerSkel :: (Word -> E.Encoding)
+                    -> (container -> Int)
+                    -> (accumFunc -> E.Encoding -> container -> E.Encoding)
+                    -> accumFunc
+                    -> container
+                    -> E.Encoding
+encodeContainerSkel encodeLen size foldr f  c =
+    encodeLen (fromIntegral (size c)) <> foldr f mempty c
+{-# INLINE encodeContainerSkel #-}
+
+decodeContainerSkelWithReplicate
+  :: Bi a
+  => D.Decoder s Int
+     -- ^ How to get the size of the container
+  -> (Int -> D.Decoder s a -> D.Decoder s container)
+     -- ^ replicateM for the container
+  -> ([container] -> container)
+     -- ^ concat for the container
+  -> D.Decoder s container
+decodeContainerSkelWithReplicate decodeLen replicateFun fromList = do
+    -- Look at how much data we have at the moment and use it as the limit for
+    -- the size of a single call to replicateFun. We don't want to use
+    -- replicateFun directly on the result of decodeLen since this might lead to
+    -- DOS attack (attacker providing a huge value for length). So if it's above
+    -- our limit, we'll do manual chunking and then combine the containers into
+    -- one.
+    size <- decodeLen
+    limit <- D.peekAvailable
```

Comment:
Accoding to [its documentation](https://github.com/well-typed/binary-serialise-cbor/blob/master/cborg/src/Codec/CBOR/Decoding.hs#L526-L532):

> Peek and return the length of the current buffer that we're running our decoder on.